File: cmd/tf-predeploy-firewall/dryrun.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/foadtalsi/tf-predeploy-firewall/internal/diff"
	"github.com/foadtalsi/tf-predeploy-firewall/internal/report"
	"github.com/foadtalsi/tf-predeploy-firewall/internal/rules"
)

// runRulesDryRun answers the question every custom-rule author has and the
// real scan can't safely answer: "what would this rule match, today, across
// the whole repo?" — without failing CI, posting comments, or reporting
// usage. Returns the process exit code.
//
// It exits 0 even when rules match: matches here are the author's feedback
// loop, not violations. The one thing that does exit non-zero is a rule
// file that doesn't load, because a config that can't parse would fail the
// real scan too, and this is the place to find that out.
func runRulesDryRun(configPath, repoDir string) int {
	custom, err := loadCustomRules(configPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "tf-predeploy-firewall: %v\n", err)
		return 2
	}
	if custom == nil {
		fmt.Fprintf(os.Stderr, "tf-predeploy-firewall: no custom_rules in %s — nothing to dry-run\n", configPath)
		return 2
	}

	files, err := diff.AllTerraformFiles(repoDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "tf-predeploy-firewall: %v\n", err)
		return 2
	}

	// Only the custom rules run: the built-ins have their own tests, and
	// mixing their findings in would bury the signal the author came for.
	// No ignores either — an author needs to see what a rule REALLY
	// matches; the real scan applies suppressions later.
	result, err := rules.Run(files, nil, []rules.Rule{custom.AsEngineRule()}, rules.RunOptions{RepoDir: repoDir})
	if err != nil {
		fmt.Fprintf(os.Stderr, "tf-predeploy-firewall: %v\n", err)
		return 2
	}

	byRule := map[string][]report.Finding{}
	for _, f := range result.Findings {
		id := strings.TrimPrefix(string(f.Category), "custom:")
		byRule[id] = append(byRule[id], f)
	}

	fmt.Printf("dry run: %d custom rule(s) against %d .tf file(s) in %s\n\n", len(custom.Rules), len(files), repoDir)
	for _, r := range custom.Rules {
		matches := byRule[r.ID]
		// "matched nothing" is the line an author most needs to see — a
		// rule that silently matches nothing is indistinguishable from a
		// working one until the incident it should have caught.
		fmt.Printf("rule %q: %d match(es)\n", r.ID, len(matches))
		sort.Slice(matches, func(i, j int) bool {
			if matches[i].File != matches[j].File {
				return matches[i].File < matches[j].File
			}
			return matches[i].Line < matches[j].Line
		})
		for _, m := range matches {
			fmt.Printf("  %s:%d  %s  [%s] %s\n", m.File, m.Line, m.Resource, m.Severity, m.Message)
		}
		fmt.Println()
	}
	return 0
}

File: src/tfpdf/cli/dryrun.py
"""`--rules-dry-run` : que trouveraient mes règles personnalisées, aujourd'hui,
sur tout le dépôt ?

Port de cmd/tf-predeploy-firewall/dryrun.go.
"""

from __future__ import annotations

import sys

from .. import diff, rules
from ..report.finding import Finding
from .config import ConfigError, load_custom_rules


def run_rules_dry_run(config_path: str, repo_dir: str) -> int:
    """Répond à la question que se pose tout auteur de règle personnalisée et à laquelle le vrai
    scan ne peut pas répondre sans risque — sans faire échouer la CI, sans poster de
    commentaires et sans rapporter d'usage. Rend le code de sortie du processus."""
    try:
        custom_rules = load_custom_rules(config_path)
    except ConfigError as exc:
        print(f"tf-predeploy-firewall: {exc}", file=sys.stderr)
        return 2
    if custom_rules is None:
        print(
            f"tf-predeploy-firewall: no custom_rules in {config_path} — nothing to dry-run",
            file=sys.stderr,
        )
        return 2

    try:
        files = diff.all_terraform_files(repo_dir)
    except Exception as exc:
        print(f"tf-predeploy-firewall: {exc}", file=sys.stderr)
        return 2

    # Only the custom rules run: the built-ins have their own tests, and mixing
    # their findings in would bury the signal the author came for. No ignores
    # either — an author needs to see what a rule REALLY matches; the real scan
    # applies suppressions later.
    try:
        result = rules.run(
            files, None, [custom_rules.as_engine_rule()], rules.RunOptions(repo_dir=repo_dir)
        )
    except Exception as exc:
        print(f"tf-predeploy-firewall: {exc}", file=sys.stderr)
        return 2

    findings_by_rule: dict[str, list[Finding]] = {}
    for finding in result.findings:
        rule_id = str(finding.category).removeprefix("custom:")
        findings_by_rule.setdefault(rule_id, []).append(finding)

    print(
        f"dry run: {len(custom_rules.rules)} custom rule(s) against {len(files)} .tf file(s) "
        f"in {repo_dir}\n"
    )
    for custom_rule in custom_rules.rules:
        matches = sorted(
            findings_by_rule.get(custom_rule.id, []),
            key=lambda matched_finding: (matched_finding.file, matched_finding.line),
        )
        # "matched nothing" is the line an author most needs to see — a rule
        # that silently matches nothing is indistinguishable from a working one
        # until the incident it should have caught.
        print(f'rule "{custom_rule.id}": {len(matches)} match(es)')
        for matched_finding in matches:
            print(
                f"  {matched_finding.file}:{matched_finding.line}  {matched_finding.resource}  [{matched_finding.severity}] {matched_finding.message}"
            )
        print()
    return 0
